### PR TITLE
Feature/issue #161/persist tokens and user information

### PR DESCRIPTION
**Pull request by JRB958**
### Login Functionality in `AuthContext`
  - Allows users to authenticate with an identifier (e.g., email or username) and password in `login` function in `AuthContext` called from `login` component.
  - Stores the received access and refresh tokens securely using `AsyncStorage` (to be encrypted in future PR). Since SecureStorage can't handle tokens of size 2048.
  - Updates the userID and isAuthenticated context state to reflect successful login across the application.
  - Resets logoutRequested to false to ensure a fresh session.
  - initiate the access token refresh handled in function `scheduleTokenRefresh`

### Logout Functionality in `AuthContext`
  - Clears access and refresh tokens from `AsyncStorage`.
  - Resets the `userID` and `isAuthenticated` states to ensure the user is logged out.
  - Redirects the user to the login screen and resets the navigation stack (had some bugs but should be good now).
  - Prevents further token refreshes by setting logoutRequested to true.
 
### Token Refresh Mechanism:
  - Decodes the access token to schedule a refresh before it expires.
  - Automatically refreshes the access token using the stored refresh token.
  - Prevents duplicate token refresh calls with proper state management and a guard variable (isTimerActive).
  - Handles cases where the refresh token is invalid or expired by logging the user out.
### Error Handling:
  - Captures errors during login, logout, and token refresh, ensuring proper error messages and logging.
  - Handles edge cases such as missing tokens, expired tokens, and network failures.
  - Improved Logging for Debugging:
  - Added detailed logs to track the flow of authentication, token management, and timer scheduling.
  - Logs capture state before and after key actions for easier debugging.
### AuthContext:
  - Provides a React context for managing authentication state and actions (login, logout, refreshAccessToken).
  - Ensures all authentication-related functionality is encapsulated in a single provider for reuse across the application.
#### Helper Functions:
  - Functions for securely storing, retrieving, and removing tokens using AsyncStorage.
  - Utility to decode JWT tokens for expiration checks.
#### State Management:
  - Uses useState for reactive state (userID, isAuthenticated).
  - Utilizes useRef for non-reactive state (logoutRequested, isTimerActive) to manage flags and prevent unnecessary re-renders.
### Manual Tests:
  - Tested login with valid.
  - Verified token storage, retrieval, and clearance in `AsyncStorage`.
  - Simulated access token expiration and confirmed successful refresh.
  - Validated proper logout behavior and prevention of further refreshes after logout.
  - Ensured no duplicate refresh calls using isTimerActive guard.
### Edge Cases Handled:
  - Access token expiration during active sessions and will automatically refresh.
  - Logout during an active token refresh or timer.
  - Login for at least 30 minutes successfully without abrupt logout
  - consecutive login-logout without an issue
  
## Remaining Tasks
  - Missing or expired refresh tokens leading to user logout.
  - handled application closure or background/foreground 
  - Test with invalid credentials response
  - Add automated tests for AuthContext using Jest and React Testing Library to validate all edge cases and scenarios.
  - Remove or minimize debug console.log statements before merging.
  - Ensure logout API endpoint integration (if applicable).
  - go over the tagged TODO and FIXME

### CODE REVIEW

**Review by NicolasChelico — DISMISSED** [review]

File: ClientApp/api/axiosInstance.ts [hunk 1]
```diff
@@ -114,47 +89,64 @@ axiosInstance.interceptors.request.use(
...
+          break;
+        case 401:
+          const newAccessToken = await refreshAccessToken();
+          if (newAccessToken) {
+            originalRequest.headers.Authorization = `Bearer ${newAccessToken}`;
+            return axiosInstance(originalRequest); // Retry original request
+          } else {
+            consoleError(ALERT_MESSAGES.sessionExpired.title, ALERT_MESSAGES.sessionExpired.message, status);
+            
+            // Logout user
```
Comment (KhalilGarali) [inline ClientApp/api/axiosInstance.ts:116]: We can encapsulate the logout logic here, maybe me centralize it somewhere else. It could be something like:
```
const handleLogout = async () => {
  try {
    await AsyncStorage.multiRemove(['accessToken', 'refreshToken']);
  } catch (error) {
    console.error('Error during logout:', error);
  }
};
```

If I am not mistaken I think you did a logout method somewhere, you can maybe use that?

File: ClientApp/api/axiosInstance.ts [hunk 2]
```diff
@@ -29,71 +29,46 @@ const config: AxiosRequestConfig = {
...
-export const setRefreshToken = async (token: string) => {
-  await AsyncStorage.setItem('refresh_token', token);
-}
-// 3- getRefreshToken function to retrieve the refresh token from AsyncStorage
-const getRefreshToken = async (): Promise<string | null> => {
-  return await AsyncStorage.getItem('refresh_token');
-};
+// setter functions to store the tokens in AsyncStorage
+export const setAccessToken = async (token: string): Promise<void> => await AsyncStorage.setItem('accessToken', token);
+export const setRefreshToken = async (token: string) => {await AsyncStorage.setItem('refresh_token', token);}
```
[the remark this answers is not in the document]
Comment (JRB958) [reply ClientApp/api/axiosInstance.ts:37]: nope, it was a typo after refactoring

File: ClientApp/api/axiosInstance.ts (hunk 1 above)
Comment (JRB958) [reply ClientApp/api/axiosInstance.ts:116]: Good note! this will be addressed soon by implementing global events

I have to think about how to do that, because the function `logout` that I have defined in `AuthContext` can't be used in the `axiosInstance` because `AuthContext` imports `axiosInstance` and If I use a function from `AuthContext` in `axiosInstance` it will lead to circular imports which will cause faulty behaviour..I need to centralize the logic somewhere else but for now this is ok for user login and logout

**Review by NicolasChelico — APPROVED** [review]

**Review by KhalilGarali — APPROVED** [review]
Good for now!
